hudi-client/hudi-spark-client/src/test/java/org/apache/hudi/table/action/commit/TestCopyOnWriteActionExecutor.java

```diff
@@ -456,4 +479,95 @@ public void testBulkInsertRecords(String bulkInsertMode) throws Exception {
+    RawTripTestPayload rowChange3 = new RawTripTestPayload(recordStr3);
+    records.add(new HoodieRecord(new HoodieKey(rowChange3.getRowKey(), rowChange3.getPartitionPath()), rowChange3));
+
+    writeClient.insert(jsc.parallelize(records, 1), firstCommitTime);
+    List<String> firstInsertFileGroupIds = table.getFileSystemView().getAllFileGroups(partitionPath)
+        .map(fileGroup -> fileGroup.getFileGroupId().getFileId()).collect(Collectors.toList());
+    List<List<FileSlice>> firstInsertFileSlicesList = table.getFileSystemView().getAllFileGroups(partitionPath)
+        .map(fileGroup -> fileGroup.getAllFileSlices().collect(Collectors.toList())).collect(Collectors.toList());
+    assertEquals(2, firstInsertFileGroupIds.size());
+    List<FileSlice>[] fileSlices = (List<FileSlice>[])firstInsertFileSlicesList.toArray(new List[firstInsertFileSlicesList.size()]);
+
+    // 2. generate clustering plan the filegroups
+    String clusterTime1 = "1";
+    createRequestedReplaceInstant(this.metaClient, clusterTime1, fileSlices);
+
+    // 3. insert one record with no updating reject exception
+    String insertRecordStr4 = "{\"_row_key\":\"8eb5b87d-1fej-4edd-87b4-6ec96dc40pp0\","
+        + "\"time\":\"2016-01-31T03:16:41.415Z\",\"number\":51}";
+    RawTripTestPayload rowChange4 = new RawTripTestPayload(insertRecordStr4);
+    HoodieRecord insertedRecord1 =
+        new HoodieRecord(new HoodieKey(rowChange4.getRowKey(), rowChange4.getPartitionPath()), rowChange4);
+    List<HoodieRecord> insertRecords = Arrays.asList(insertedRecord1);
+    Thread.sleep(1000);
+    String secondCommitTime = makeNewCommitTime();
+    metaClient = HoodieTableMetaClient.reload(metaClient);
+    writeClient.startCommitWithTime(secondCommitTime);
+    List<WriteStatus> statuses = writeClient.upsert(jsc.parallelize(insertRecords), secondCommitTime).collect();
+    assertEquals(1, statuses.size());
+
+    // 4. insert one record and update one record
+    String insertRecordStr5 = "{\"_row_key\":\"8eb5b87d-1fej-4edd-87b4-6ec96dc405a0\","
+        + "\"time\":\"2016-01-31T03:16:41.415Z\",\"number\":51}";
+    // We update the 1st record & add a new record
+    String updateRecordStr1 = "{\"_row_key\":\"8eb5b87a-1feh-4edd-87b4-6ec96dc405a0\","
+        + "\"time\":\"2016-01-31T03:16:41.415Z\",\"number\":15}";
+    RawTripTestPayload updateRowChanges1 = new RawTripTestPayload(updateRecordStr1);
+    HoodieRecord updatedRecord1 = new HoodieRecord(
+        new HoodieKey(updateRowChanges1.getRowKey(), updateRowChanges1.getPartitionPath()), updateRowChanges1);
+    RawTripTestPayload rowChange5 = new RawTripTestPayload(insertRecordStr5);
+    HoodieRecord insertedRecord2 =
+        new HoodieRecord(new HoodieKey(rowChange5.getRowKey(), rowChange5.getPartitionPath()), rowChange5);
+    List<HoodieRecord> updatedRecords = Arrays.asList(updatedRecord1, insertedRecord2);
+    Thread.sleep(1000);
```

Comment:
please avoid sleep in tests. It tends to cause flakiness. If we can write the test such that it waits til a certain condition is met (or times out), that would be much more preferable.

Reply:
okay